Source/Task/LocklessQueue.h

```diff
@@ -0,0 +1,740 @@
+        }
+        else
+        {
+            address = 0;
+            return false;
+        }
+    }
+
+    //
+    // Frees a previously reserved node.  Freeing a node
+    // that has been pushed results in undefined behavior. It
+    // will be hard to debug.
+    //
+    void free_node(_In_ uint64_t address) noexcept
+    {
+        Address a;
+        a.value = address;
+        Node* n = to_node(a);
+        m_heap.free(n, a);
+    }
+    
+    //
+    // Pushes the given data onto the back
+    // of the queue.  A copy is made of TData.
+    // If the push fails this returns false.
+    //
+    bool push_back(_In_ TData data) noexcept
+    {
+        Address address;
+        Node* node = m_heap.alloc(address);
+        
+        if (node != nullptr)
+        {
+            node->data = std::move(data);
+            m_activeList.push(node, address);
+            return true;
+        }
+        
+        return false;
+    }
+
+    //
+    // Pushes the given data onto the back of the queue,
+    // using a reserved node pointer.  This never fails as
+    // the node pointer was preallocated.
+    //
+    void push_back(_In_ TData data, _In_ uint64_t address)
+    {
+        Address a;
+        a.value = address;
+        Node* n = to_node(a);
+        n->data = std::move(data);
+        m_activeList.push(n, a);
+    }
+    
+    //
+    // Pops TData off the head of the queue, returning
+    // true if successful or false if the queue is
+    // empty.
+    //
+    bool pop_front(_Out_ TData& data) noexcept
+    {
+        Address address;
+        Node* node = m_activeList.pop(address);
+        
+        if (node != nullptr)
+        {
+            data = std::move(node->data);
+            node->data = TData {};
+            m_heap.free(node, address);
+            return true;
+        }
+        
+        return false;
+    }
+
+    //
+    // Pops TData off the front of the queue, returning
+    // true if successful, and returns a node address
+    // instead of placing the address back on the free
+    // list. This address can later be used in a push_back
+    // call or a free_node call.  Failing to call either of these
+    // results in a leaked node. Returns false if the queue
+    // is empty.
+    //
+    bool pop_front(_Out_ TData& data, _Out_ uint64_t& address) noexcept
+    {
+        Address a;
+        Node* node = m_activeList.pop(a);
+        
+        if (node != nullptr)
+        {
+            data = std::move(node->data);
+            node->data = TData {};
+            address = a.value;
+            return true;
+        }
+        
+        return false;
+    }
+
+private:
+
+    /*
+     *
+     * Structure Definitions
+     *
+     */
+    
+    // Address - This represents the address of a node.
+    // Nodes live in a contiguous memory block, and there are multiple
+    // blocks. Address represents the position of the
+    // node and must be 64 bits.
+    struct Address
+    {
+        union
```

Comment:
type punning through a union is technically undefined behavior (breaks strict aliasing I think)

Reply:
an alternative would be to have accessors for index/block/aba that do the appropriate bit masking and shifting

Reply:
Would be slower.  I've read a little more on this. I get why this is bad for complex types, and ambiguous for passing pointers around.  But all the "cleaner" ways of doing this I've seen aren't...cleaner.  Best I've seen is a using memcpy to blit the data out.  Looks terrible in code but apparently memcpy is so well optimized by modern compilers that it folds down to a single instruction if the data size fits in a register.